src/Datadog.Trace.ClrProfiler.Managed/Integrations/RabbitMQIntegration.cs

```diff
@@ -0,0 +1,910 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Datadog.Trace.ClrProfiler.Emit;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Tagging;
+
+namespace Datadog.Trace.ClrProfiler.Integrations
+{
+    /// <summary>
+    /// Tracing integration for RabbitMQ.Client
+    /// </summary>
+    public static class RabbitMQIntegration
+    {
+        internal const string IntegrationName = nameof(IntegrationIds.RabbitMQ);
+
+        private const string OperationName = "amqp.command";
+        private const string ServiceName = "rabbitmq";
+
+        private const string Major3Minor6Patch9 = "3.6.9";
+        private const string Major5 = "5";
+        private const string Major6 = "6";
+        private const string RabbitMQAssembly = "RabbitMQ.Client";
+        private const string RabbitMQIModel = "RabbitMQ.Client.IModel";
+        private const string RabbitMQImplModelBase = "RabbitMQ.Client.Impl.ModelBase";
+        private const string RabbitMQDefaultBasicConsumer = "RabbitMQ.Client.DefaultBasicConsumer";
+        private const string IBasicPropertiesTypeName = "RabbitMQ.Client.IBasicProperties";
+        private const string IDictionaryArgumentsTypeName = "System.Collections.Generic.IDictionary`2[System.String,System.Object]";
+
+        internal static readonly IntegrationInfo IntegrationId = IntegrationRegistry.GetIntegrationInfo(nameof(IntegrationIds.RabbitMQ));
+        private static readonly Vendors.Serilog.ILogger Log = DatadogLogging.GetLogger(typeof(RabbitMQIntegration));
+        private static readonly string[] DeliveryModeStrings = { null, "1", "2" };
+
+        private static Action<IDictionary<string, object>, string, string> headersSetter = ((carrier, key, value) =>
+        {
+            carrier[key] = Encoding.UTF8.GetBytes(value);
```

Comment:
Let me know if this is something I need to address. I started off with just a `Dictionary.Add` because we should only be doing this once per `BasicProperties.Headers`. However, there's differing behaviors between library versions of the client library when `BasicProperties` is null. In older versions, a brand new object is created (a blank slate if you will). In newer versions, an "empty" Singleton is created and re-used for all invocations where it is null. Thus, when re-used the provided `BasicProperties` object already has Datadog-specific values added to the Dictionary, causing `Dictionary.Add` to fail. I fixed it by doing this hard assignment. Is this fine or do you think I should change my approach somehow?

Reply:
The only issue I see with the current approach is that we execute `UTF8.GetBytes(value)` and allocate the byte array even if the header is already present.

🧠⚡ One option is to call `Contains(key)` first to avoid the allocation if not needed, but we should probably benchmark that to make sure we're not making things worse with the additional dictionary lookup.

Reply:
We'll need to override it anyways because if the object was cached, the header may already be present and its value may be invalid so regardless we need to set the value

Reply:
Oh, never mind then!

Reply:
This is a good idea for an integration test though...thanks for giving me the idea!

Reply:
I added another assertion in 2309405f66c1421a1ea9218b996c3f8a5c6d3ddf to ensure that each publish adds its span context to the message. Locally I tested the scenario of bailing out if `Contains(key)`, but because of the cached properties this led to some publishes reusing the previous span context, leading to one `publish` span being the parent for 3 `deliver` spans, which is wrong. So this will be a good assert to keep around 😄